File: src/mod3_oop/lesson38_functional_interface/OOPDemo.java
package mod3_oop.lesson38_functional_interface;

public class OOPDemo {
    public static void main(String[] args) {
        // with classes which implements Greeting interface
//        Greeting englishGreeting = new EnglishGreeting();
//        englishGreeting.sayHello();
//
//        Greeting frenchGreeting = new FrenchGreeting();
//        frenchGreeting.sayHello();

        Greeting englishGreeting = name -> System.out.println("Hello, " + name + "!");
        englishGreeting.sayHello("Kostya");

        new Greeting() {
            @Override
            public void sayHello(String name) {
                System.out.println("Привет " + name + "!");
            }
        }.sayHello("Костя");

        Greeting frenchGreeting = name -> System.out.println("Bonjour, " + name + "!");
        frenchGreeting.sayHello("Tolya");
    }
}

File: src/mod3_oop/lesson38_functional_interface/OOPDemo.kt
package mod3_oop.lesson38_functional_interface

fun main() {
    val englishGreeting: (String) -> Unit = { name -> println("Hello, $name!")}
    englishGreeting("Kostya")

    object : GreetingKt {
        override fun sayHello(name: String) {
            println("Bonjour, $name!")
        }
    }.sayHello("Fransua Mean")

    val russianGreeting = GreetingKt {name: String -> println("Привет, $name!")}
    russianGreeting.sayHello("Костя")
}